### PR TITLE
Fix regression tests and start supporting more expression types.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is only relevant for the last 3 commits. WIll be rebased once the dependent PR (https://github.com/diffblue/gnat2goto/pull/25) gets merged.

### CODE REVIEW

File: testsuite/gnat2goto/tests/loops/loops.adb [hunk 1]
```diff
@@ -44,4 +44,9 @@ begin
    while z > 1 loop
       exit;
    end loop;
+
+   for i in 1..10 loop
+     pragma assert(i < 10);
+     null;
+   end loop;
```
Comment (hannes-steffenhagen-diffblue) [inline testsuite/gnat2goto/tests/loops/loops.adb]: This part was meant to test loop unwinding, I don't think it's really necessary anymore.

File: testsuite/gnat2goto/tests/long_integer/test_long_integer.adb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+with Ada.Text_IO;
+procedure Test_Long_Integer is
+  x: Long_Long_Integer := 1152921504606846976;
```
Comment (smowton) [inline testsuite/gnat2goto/tests/long_integer/test_long_integer.adb]: Check: does cbmc require widths to be multiples of a byte? Does Ada?

File: gnat2goto/driver/tree_walk.adb [hunk 3]
```diff
@@ -1206,12 +1209,24 @@ package body Tree_Walk is
...
+            when N_And_Then => Do_And_Then (N),
             when N_Or_Else => Create_Dummy_Irep,
             when N_Qualified_Expression => Create_Dummy_Irep,
             when others                 => raise Program_Error);
    end Do_Expression;
 
+   function Do_And_Then (N : Node_Id) return Irep is
+      L : constant Node_Id := Left_Opnd (N);
+      R : constant Node_Id := Right_Opnd (N);
+      Expr : constant Irep := New_Irep (I_Op_And);
```
Comment (smowton) [inline gnat2goto/driver/tree_walk.adb]: Check: this should be `&&` (short-circuit boolean) not `&`

File: testsuite/gnat2goto/tests/long_integer/test_long_integer.adb (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply testsuite/gnat2goto/tests/long_integer/test_long_integer.adb]: W.r.t Ada: No, consider

```
with Ada.Text_IO;
procedure Size_Test is

  package IO renames Ada.Text_IO;

  type My_Bit is (On, Off)
    with Size => 1;

  type My_Bits is Array(Integer range <>) of My_Bit
    with Pack;
  
  Bit : My_Bit := Off;

  type My_Rec is record
    Bits : My_Bits (0..7);
  end record;
begin
  IO.Put_Line ("Bit: " & My_Bit'Image (Bit));
  Bit := On;
  IO.Put_Line ("Bit: " & My_Bit'Image (Bit));
  IO.Put_Line ("Bit size: " & Integer'Image (My_Bit'Size));
  IO.Put_Line ("Struct size: " & Integer'Image (My_Rec'Size));
end Size_Test;
```

Output will be

> Bit: OFF
Bit: ON
Bit size:  1
Struct size:  8

As to cbmc, I don't think so? We already support bitfields which require odd sizes, and `bitvector_typet` - which is the base of most primitives - doesn't seem to place any restrictions on the width passed (which is in bits).

File: gnat2goto/driver/tree_walk.adb (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply gnat2goto/driver/tree_walk.adb]: `I_Op_And` maps to `"and"` in `ireps.adb`, which is the same as `ID_and` in cbmc which is `&&`. `&` is `ID_bitand` in cbmc.

**Pull request merged**